File: 05_NumberTheory/A4-NextPrime/A4.py
class Primos:
    def __init__(self, n):
        self.lista_primos = []  # lista de números primos
        self.min_primo = []  # menor primo que divide a i
        self.es_primo = []  # True si es primo
        self.criba(n)
    
    def criba(self, n):
        self.min_primo = [0] * (n + 1)
        self.es_primo = [True] * (n + 1)  # En principio son todos primos
        self.es_primo[0] = self.es_primo[1] = False  # El 0 y 1 no son primos
        
        for i in range(2, n + 1):
            if not self.es_primo[i]:
                continue
            self.lista_primos.append(i)
            j = i * i
            while j <= n:
                if self.es_primo[j]:
                    self.es_primo[j] = False
                    self.min_primo[j] = i
                    # i es el menor primo que divide a j
                j += i
    
    def test_primalidad(self, x):
        if x < len(self.es_primo):
            return self.es_primo[x]
        
        # Asegurarse que estén hallados todos los primos
        # menores o igual a sqrt(x) en la criba
        for p in self.lista_primos:
            if x % p == 0:
                return False
        return True


def main():
    primos = Primos(int(1e6) + 1000)
    t = int(input())
    
    for _ in range(t):
        x = int(input())
        i = x + 1
        while True:
            ans = primos.test_primalidad(i)
            if ans:
                print(i)
                break
            i += 1


if __name__ == "__main__":
    main()

File: 05_NumberTheory/A4-NextPrime/A4.cpp
#include <bits/stdc++.h>

using namespace std;
#define FIN ios::sync_with_stdio(0);cout.tie(0);cin.tie(0)

struct Primos {
    typedef long long ll;
    vector<int> listaPrimos; // lista de numeros primos
    vector<int> minPrimo; // menor primo que divide a i
    // Saber el menor primo que divide a un numero nos permite
    // factorizar mas rapido
    vector<bool> esPrimo; // True si es primo
    Primos(int n) {criba(n);}
    void criba(int n) {
        minPrimo.resize(n+1);
        esPrimo.resize(n+1,true); // En principio son todos primos
        esPrimo[0] = esPrimo[1] = false; // El 1 no es primo
        for(int i = 2; i <= n; i++) {
            if(!esPrimo[i]) continue;
            listaPrimos.push_back(i);
            for(ll j = 1LL*i*i; j <= n; j += i) {
                if(esPrimo[j]) {
                    esPrimo[j] = false;
                    minPrimo[j] = i; 
                    // i es el menor primo que divide a j
                }
            }
        }
    }
    bool testPrimalidad(ll x) {
        if(x < esPrimo.size()) return esPrimo[x];
        ll maxPrimo = listaPrimos.back();
        // Asegurarse que esten hallados todos los primos
        // menores o igual a sqrt(x) en la criba
        for(int p : listaPrimos) {
            if(x%p == 0) return false;
        }
        return true;
    }
};

typedef long long ll;

int main() {
    FIN;

    Primos primos(1e6+1000);
    int t; cin >> t;
    while(t--) {
        ll x; cin >> x;
        for(ll i = x+1; ;i++) {
            bool ans = primos.testPrimalidad(i);
            if(ans) {
                cout << i << "\n";
                break;
            } 
        }
    }

    return 0;
}